### PR TITLE
Lua: add darkroom-image-loaded event and clear_snapshots() function

### PR DESCRIPTION
**Pull request by wpferguson**
Added "darkroom-image-loaded event that fires when an image is loaded into darkroom mode.  The image is returned to the event handler when the event fires.  Needed to build a solution to #2315.  Also needed (along with some other events that don't exist yet) to implement responsive cache updating (lua script to keep mipmap cache up to date by catching image changes and updating the cache in the lua thread).

Added function clear_snapshots() to remove the list of snapshots.  While testing the script to automatically generate a snapshot when an image was loaded in darkroom, I ended up with a string of snapshots named "exposure(11)".  The newest was always at the top, but for this function having that list didn't really make sense so I added a function to clear the list before taking the snapshot.

To test:  add the attached test_darkroom_trigger.lua script to your scripts and enable it.  Open an image in darkroom view and it will automatically create a snapshot.  Each time you change images, the snapshot list will be cleared and a new snapshot created.

### CODE REVIEW

**Comment by wpferguson** [comment]
Oops, forgot to attach the script

[test_darkroom_trigger.zip](https://github.com/darktable-org/darktable/files/6985708/test_darkroom_trigger.zip)

**Comment by ptilopteri** [comment]
fails to load for me
70.868654 LUA ERROR: script_manager.lua: activate: 345: Error loading contrib/test_darkroom_trigger 
70.868826 LUA ERROR: script_manager.lua: activate: 346: Error message: .../.config/darktable/lua/contrib/test_darkroom_trigger.lua:3: unknown event type : darkroom-image-loaded

g6f6fff41b
on openSUSE Tumbleweed 20210810

**Comment by wpferguson** [comment]
@ptilopteri did you download the branch with the source code changes and compile them in?

**Comment by ptilopteri** [comment]
no, dl the zip you provided, extracted test_darkroom_trigger.lua and 
placed into ~/.config/darktable/lua/contrib

**Comment by wpferguson** [comment]
The script requires the code changes to master be compiled in so that the event, "darkroom-image-loaded", is available to use.  The script was provided to test the code changes.

**Comment by johnny-bit** [comment]
(nitpicking but you've got `clear_shortcuts()` in title while it should be `clear_snapshots()` :P)

File: src/libs/snapshots.c [hunk 1]
```diff
@@ -668,6 +683,21 @@ static int lua_take_snapshot(lua_State *L)
   return 0;
 }
 
+static int lua_clear_snapshots(lua_State *L)
+{
+  dt_lib_module_t *self = lua_touserdata(L, lua_upvalueindex(1));
+  dt_lib_snapshots_t *d = (dt_lib_snapshots_t *)self->data;
```
Comment (TurboGit) [inline src/libs/snapshots.c:689]: from here, why not call `_internal_gui_reset(self);` The only diff is the redraw center, but well, maybe its needed when called from Lua ?

Comment (wpferguson) [reply src/libs/snapshots.c:689]: Oops, _internal_gui_reset shouldn't be there.  My first cut at it was copy `gui_reset` to `_internal_gui_reset` and calling `_internal_gui_reset` from `gui_reset` and `clear_snapshots`.  When I tested it appeared that I got a screen flash when I loaded the image and I figured it was the `dt_control_queue_redraw_center` causing it.  Let me do some more testing and see what I can come up with.

Comment (wpferguson) [reply src/libs/snapshots.c]: Got rid of `_internal_gui_reset`.  Changed `lua_clear_snapshots` to just use `gui_reset`.  Tested and didn't see any screen flashes.

**Review by TurboGit — APPROVED** [review]
Looks good, thanks !

**Pull request merged**